File: flyteadmin/pkg/repositories/transformers/node_execution.go
package transformers

import (
	"context"

	"github.com/golang/protobuf/proto"
	"github.com/golang/protobuf/ptypes"
	"google.golang.org/grpc/codes"

	"github.com/flyteorg/flyte/flyteadmin/pkg/common"
	"github.com/flyteorg/flyte/flyteadmin/pkg/errors"
	genModel "github.com/flyteorg/flyte/flyteadmin/pkg/repositories/gen/models"
	"github.com/flyteorg/flyte/flyteadmin/pkg/repositories/models"
	"github.com/flyteorg/flyte/flyteadmin/pkg/runtime/interfaces"
	"github.com/flyteorg/flyte/flyteidl/gen/pb-go/flyteidl/admin"
	"github.com/flyteorg/flyte/flyteidl/gen/pb-go/flyteidl/core"
	"github.com/flyteorg/flyte/flyteidl/gen/pb-go/flyteidl/event"
	"github.com/flyteorg/flyte/flytestdlib/logger"
	"github.com/flyteorg/flyte/flytestdlib/storage"
)

type ToNodeExecutionModelInput struct {
	Request                      *admin.NodeExecutionEventRequest
	ParentTaskExecutionID        *uint
	ParentID                     *uint
	DynamicWorkflowRemoteClosure string
	InlineEventDataPolicy        interfaces.InlineEventDataPolicy
	StorageClient                *storage.DataStore
}

func addNodeRunningState(request *admin.NodeExecutionEventRequest, nodeExecutionModel *models.NodeExecution,
	closure *admin.NodeExecutionClosure) error {
	occurredAt, err := ptypes.Timestamp(request.GetEvent().GetOccurredAt())
	if err != nil {
		return errors.NewFlyteAdminErrorf(codes.Internal, "failed to unmarshal occurredAt with error: %v", err)
	}

	nodeExecutionModel.StartedAt = &occurredAt
	startedAtProto, err := ptypes.TimestampProto(occurredAt)
	if err != nil {
		return errors.NewFlyteAdminErrorf(codes.Internal,
			"failed to marshal occurredAt into a timestamp proto with error: %v", err)
	}
	closure.StartedAt = startedAtProto
	closure.DeckUri = request.Event.DeckUri
	return nil
}

func addTerminalState(
	ctx context.Context,
	request *admin.NodeExecutionEventRequest, nodeExecutionModel *models.NodeExecution,
	closure *admin.NodeExecutionClosure, inlineEventDataPolicy interfaces.InlineEventDataPolicy, storageClient *storage.DataStore) error {
	if closure.GetStartedAt() == nil {
		logger.Warning(context.Background(), "node execution is missing StartedAt")
	} else {
		endTime, err := ptypes.Timestamp(request.GetEvent().GetOccurredAt())
		if err != nil {
			return errors.NewFlyteAdminErrorf(
				codes.Internal, "Failed to parse node execution occurred at timestamp: %v", err)
		}
		nodeExecutionModel.Duration = endTime.Sub(*nodeExecutionModel.StartedAt)
		closure.Duration = ptypes.DurationProto(nodeExecutionModel.Duration)
	}

	// Serialize output results (if they exist)
	if request.GetEvent().GetOutputUri() != "" {
		closure.OutputResult = &admin.NodeExecutionClosure_OutputUri{
			OutputUri: request.GetEvent().GetOutputUri(),
		}
	} else if request.GetEvent().GetOutputData() != nil {
		switch inlineEventDataPolicy {
		case interfaces.InlineEventDataPolicyStoreInline:
			closure.OutputResult = &admin.NodeExecutionClosure_OutputData{
				OutputData: request.GetEvent().GetOutputData(),
			}
		default:
			logger.Debugf(ctx, "Offloading outputs per InlineEventDataPolicy")
			uri, err := common.OffloadLiteralMap(ctx, storageClient, request.GetEvent().GetOutputData(),
				request.GetEvent().GetId().GetExecutionId().GetProject(), request.GetEvent().GetId().GetExecutionId().GetDomain(), request.GetEvent().GetId().GetExecutionId().GetName(),
				request.GetEvent().GetId().GetNodeId(), OutputsObjectSuffix)
			if err != nil {
				return err
			}
			closure.OutputResult = &admin.NodeExecutionClosure_OutputUri{
				OutputUri: uri.String(),
			}
		}
	} else if request.GetEvent().GetError() != nil {
		closure.OutputResult = &admin.NodeExecutionClosure_Error{
			Error: request.GetEvent().GetError(),
		}
		k := request.GetEvent().GetError().GetKind().String()
		nodeExecutionModel.ErrorKind = &k
		nodeExecutionModel.ErrorCode = &request.Event.GetError().Code
	}
	closure.DeckUri = request.GetEvent().GetDeckUri()

	return nil
}

func CreateNodeExecutionModel(ctx context.Context, input ToNodeExecutionModelInput) (*models.NodeExecution, error) {
	nodeExecution := &models.NodeExecution{
		NodeExecutionKey: models.NodeExecutionKey{
			NodeID: input.Request.GetEvent().GetId().GetNodeId(),
			ExecutionKey: models.ExecutionKey{
				Project: input.Request.GetEvent().GetId().GetExecutionId().GetProject(),
				Domain:  input.Request.GetEvent().GetId().GetExecutionId().GetDomain(),
				Name:    input.Request.GetEvent().GetId().GetExecutionId().GetName(),
			},
		},
		Phase: input.Request.GetEvent().GetPhase().String(),
	}

	reportedAt := input.Request.GetEvent().GetReportedAt()
	if reportedAt == nil || (reportedAt.GetSeconds() == 0 && reportedAt.GetNanos() == 0) {
		reportedAt = input.Request.GetEvent().GetOccurredAt()
	}

	closure := admin.NodeExecutionClosure{
		Phase:     input.Request.GetEvent().GetPhase(),
		CreatedAt: input.Request.GetEvent().GetOccurredAt(),
		UpdatedAt: reportedAt,
	}

	nodeExecutionMetadata := admin.NodeExecutionMetaData{
		RetryGroup:   input.Request.GetEvent().GetRetryGroup(),
		SpecNodeId:   input.Request.GetEvent().GetSpecNodeId(),
		IsParentNode: input.Request.GetEvent().GetIsParent(),
		IsDynamic:    input.Request.GetEvent().GetIsDynamic(),
		IsArray:      input.Request.GetEvent().GetIsArray(),
	}
	err := handleNodeExecutionInputs(ctx, nodeExecution, input.Request, input.StorageClient)
	if err != nil {
		return nil, err
	}

	if input.Request.GetEvent().GetPhase() == core.NodeExecution_RUNNING {
		err := addNodeRunningState(input.Request, nodeExecution, &closure)
		if err != nil {
			return nil, err
		}
	}

	if common.IsNodeExecutionTerminal(input.Request.GetEvent().GetPhase()) {
		err := addTerminalState(ctx, input.Request, nodeExecution, &closure, input.InlineEventDataPolicy, input.StorageClient)
		if err != nil {
			return nil, err
		}
	}

	// Update TaskNodeMetadata, which includes caching information today.
	if input.Request.GetEvent().GetTaskNodeMetadata() != nil {
		targetMetadata := &admin.NodeExecutionClosure_TaskNodeMetadata{
			TaskNodeMetadata: &admin.TaskNodeMetadata{
				CheckpointUri: input.Request.GetEvent().GetTaskNodeMetadata().GetCheckpointUri(),
			},
		}
		if input.Request.GetEvent().GetTaskNodeMetadata().GetCatalogKey() != nil {
			st := input.Request.GetEvent().GetTaskNodeMetadata().GetCacheStatus().String()
			targetMetadata.TaskNodeMetadata.CacheStatus = input.Request.GetEvent().GetTaskNodeMetadata().GetCacheStatus()
			targetMetadata.TaskNodeMetadata.CatalogKey = input.Request.GetEvent().GetTaskNodeMetadata().GetCatalogKey()
			nodeExecution.CacheStatus = &st
		}
		closure.TargetMetadata = targetMetadata
	}

	marshaledClosure, err := proto.Marshal(&closure)
	if err != nil {
		return nil, errors.NewFlyteAdminErrorf(
			codes.Internal, "failed to marshal node execution closure with error: %v", err)
	}
	marshaledNodeExecutionMetadata, err := proto.Marshal(&nodeExecutionMetadata)
	if err != nil {
		return nil, errors.NewFlyteAdminErrorf(
			codes.Internal, "failed to marshal node execution metadata with error: %v", err)
	}
	nodeExecution.Closure = marshaledClosure
	nodeExecution.NodeExecutionMetadata = marshaledNodeExecutionMetadata
	nodeExecutionCreatedAt, err := ptypes.Timestamp(input.Request.GetEvent().GetOccurredAt())
	if err != nil {
		return nil, errors.NewFlyteAdminErrorf(codes.Internal, "failed to read event timestamp")
	}
	nodeExecution.NodeExecutionCreatedAt = &nodeExecutionCreatedAt
	nodeExecutionUpdatedAt, err := ptypes.Timestamp(reportedAt)
	if err != nil {
		return nil, errors.NewFlyteAdminErrorf(codes.Internal, "failed to read event reported_at timestamp")
	}
	nodeExecution.NodeExecutionUpdatedAt = &nodeExecutionUpdatedAt
	if input.Request.GetEvent().GetParentTaskMetadata() != nil {
		nodeExecution.ParentTaskExecutionID = input.ParentTaskExecutionID
	}
	nodeExecution.ParentID = input.ParentID
	nodeExecution.DynamicWorkflowRemoteClosureReference = input.DynamicWorkflowRemoteClosure

	internalData := &genModel.NodeExecutionInternalData{
		EventVersion: input.Request.GetEvent().GetEventVersion(),
	}
	internalDataBytes, err := proto.Marshal(internalData)
	if err != nil {
		return nil, errors.NewFlyteAdminErrorf(codes.Internal, "failed to marshal node execution data with err: %v", err)
	}
	nodeExecution.InternalData = internalDataBytes
	return nodeExecution, nil
}

func UpdateNodeExecutionModel(
	ctx context.Context, request *admin.NodeExecutionEventRequest, nodeExecutionModel *models.NodeExecution,
	targetExecution *core.WorkflowExecutionIdentifier, dynamicWorkflowRemoteClosure string,
	inlineEventDataPolicy interfaces.InlineEventDataPolicy, storageClient *storage.DataStore) error {
	err := handleNodeExecutionInputs(ctx, nodeExecutionModel, request, storageClient)
	if err != nil {
		return err
	}
	var nodeExecutionClosure admin.NodeExecutionClosure
	err = proto.Unmarshal(nodeExecutionModel.Closure, &nodeExecutionClosure)
	if err != nil {
		return errors.NewFlyteAdminErrorf(codes.Internal,
			"failed to unmarshal node execution closure with error: %+v", err)
	}
	nodeExecutionModel.Phase = request.GetEvent().GetPhase().String()
	nodeExecutionClosure.Phase = request.GetEvent().GetPhase()
	reportedAt := request.GetEvent().GetReportedAt()
	if reportedAt == nil || (reportedAt.GetSeconds() == 0 && reportedAt.GetNanos() == 0) {
		reportedAt = request.GetEvent().GetOccurredAt()
	}
	nodeExecutionClosure.UpdatedAt = reportedAt

	if request.GetEvent().GetPhase() == core.NodeExecution_RUNNING {
		err := addNodeRunningState(request, nodeExecutionModel, &nodeExecutionClosure)
		if err != nil {
			return err
		}
	}
	if common.IsNodeExecutionTerminal(request.GetEvent().GetPhase()) {
		err := addTerminalState(ctx, request, nodeExecutionModel, &nodeExecutionClosure, inlineEventDataPolicy, storageClient)
		if err != nil {
			return err
		}
	}

	// If the node execution kicked off a workflow execution update the closure if it wasn't set
	if targetExecution != nil && nodeExecutionClosure.GetWorkflowNodeMetadata() == nil {
		nodeExecutionClosure.TargetMetadata = &admin.NodeExecutionClosure_WorkflowNodeMetadata{
			WorkflowNodeMetadata: &admin.WorkflowNodeMetadata{
				ExecutionId: targetExecution,
			},
		}
	}

	// Update TaskNodeMetadata, which includes caching information today.
	if request.GetEvent().GetTaskNodeMetadata() != nil {
		targetMetadata := &admin.NodeExecutionClosure_TaskNodeMetadata{
			TaskNodeMetadata: &admin.TaskNodeMetadata{
				CheckpointUri: request.GetEvent().GetTaskNodeMetadata().GetCheckpointUri(),
			},
		}
		if request.GetEvent().GetTaskNodeMetadata().GetCatalogKey() != nil {
			st := request.GetEvent().GetTaskNodeMetadata().GetCacheStatus().String()
			targetMetadata.TaskNodeMetadata.CacheStatus = request.GetEvent().GetTaskNodeMetadata().GetCacheStatus()
			targetMetadata.TaskNodeMetadata.CatalogKey = request.GetEvent().GetTaskNodeMetadata().GetCatalogKey()
			nodeExecutionModel.CacheStatus = &st
		}
		nodeExecutionClosure.TargetMetadata = targetMetadata

		// if this is a dynamic task then maintain the DynamicJobSpecUri
		dynamicWorkflowMetadata := request.GetEvent().GetTaskNodeMetadata().GetDynamicWorkflow()
		if dynamicWorkflowMetadata != nil && len(dynamicWorkflowMetadata.GetDynamicJobSpecUri()) > 0 {
			nodeExecutionClosure.DynamicJobSpecUri = dynamicWorkflowMetadata.GetDynamicJobSpecUri()
		}
	}

	marshaledClosure, err := proto.Marshal(&nodeExecutionClosure)
	if err != nil {
		return errors.NewFlyteAdminErrorf(
			codes.Internal, "failed to marshal node execution closure with error: %v", err)
	}

	nodeExecutionModel.Closure = marshaledClosure
	updatedAt, err := ptypes.Timestamp(reportedAt)
	if err != nil {
		return errors.NewFlyteAdminErrorf(codes.Internal, "failed to parse updated at timestamp")
	}
	nodeExecutionModel.NodeExecutionUpdatedAt = &updatedAt
	nodeExecutionModel.DynamicWorkflowRemoteClosureReference = dynamicWorkflowRemoteClosure

	// In the case of dynamic nodes reporting DYNAMIC_RUNNING, the IsParent and IsDynamic bits will be set for this event.
	// Update the node execution metadata accordingly.
	if request.GetEvent().GetIsParent() || request.GetEvent().GetIsDynamic() || request.GetEvent().GetIsArray() {
		var nodeExecutionMetadata admin.NodeExecutionMetaData
		if len(nodeExecutionModel.NodeExecutionMetadata) > 0 {
			if err := proto.Unmarshal(nodeExecutionModel.NodeExecutionMetadata, &nodeExecutionMetadata); err != nil {
				return errors.NewFlyteAdminErrorf(codes.Internal,
					"failed to unmarshal node execution metadata with error: %+v", err)
			}
		}
		// Not every event sends IsParent and IsDynamic as an artifact of how propeller handles dynamic nodes.
		// Only explicitly set the fields, when they're set in the event itself.
		if request.GetEvent().GetIsParent() {
			nodeExecutionMetadata.IsParentNode = true
		}
		if request.GetEvent().GetIsDynamic() {
			nodeExecutionMetadata.IsDynamic = true
		}
		if request.GetEvent().GetIsArray() {
			nodeExecutionMetadata.IsArray = true
		}
		nodeExecMetadataBytes, err := proto.Marshal(&nodeExecutionMetadata)
		if err != nil {
			return errors.NewFlyteAdminErrorf(codes.Internal,
				"failed to marshal node execution metadata with error: %+v", err)
		}
		nodeExecutionModel.NodeExecutionMetadata = nodeExecMetadataBytes
	}

	return nil
}

func FromNodeExecutionModel(nodeExecutionModel models.NodeExecution, opts *ExecutionTransformerOptions) (*admin.NodeExecution, error) {
	var closure admin.NodeExecutionClosure
	err := proto.Unmarshal(nodeExecutionModel.Closure, &closure)
	if err != nil {
		return nil, errors.NewFlyteAdminErrorf(codes.Internal, "failed to unmarshal closure")
	}

	if closure.GetError() != nil && opts != nil && opts.TrimErrorMessage && len(closure.GetError().GetMessage()) > 0 {
		trimmedErrOutputResult := closure.GetError()
		trimmedErrMessage := TrimErrorMessage(trimmedErrOutputResult.GetMessage())
		trimmedErrOutputResult.Message = trimmedErrMessage
		closure.OutputResult = &admin.NodeExecutionClosure_Error{
			Error: trimmedErrOutputResult,
		}
	}

	var nodeExecutionMetadata admin.NodeExecutionMetaData
	err = proto.Unmarshal(nodeExecutionModel.NodeExecutionMetadata, &nodeExecutionMetadata)
	if err != nil {
		return nil, errors.NewFlyteAdminErrorf(codes.Internal, "failed to unmarshal nodeExecutionMetadata")
	}
	// TODO: delete this block and references to preloading child node executions no earlier than Q3 2022
	// This is required for historical reasons because propeller did not always send IsParent or IsDynamic in events.
	if !(nodeExecutionMetadata.GetIsParentNode() || nodeExecutionMetadata.GetIsDynamic()) {
		if len(nodeExecutionModel.ChildNodeExecutions) > 0 {
			nodeExecutionMetadata.IsParentNode = true
			if len(nodeExecutionModel.DynamicWorkflowRemoteClosureReference) > 0 {
				nodeExecutionMetadata.IsDynamic = true
			}
		}
	}

	return &admin.NodeExecution{
		Id: &core.NodeExecutionIdentifier{
			NodeId: nodeExecutionModel.NodeID,
			ExecutionId: &core.WorkflowExecutionIdentifier{
				Project: nodeExecutionModel.NodeExecutionKey.ExecutionKey.Project,
				Domain:  nodeExecutionModel.NodeExecutionKey.ExecutionKey.Domain,
				Name:    nodeExecutionModel.NodeExecutionKey.ExecutionKey.Name,
			},
		},
		InputUri: nodeExecutionModel.InputURI,
		Closure:  &closure,
		Metadata: &nodeExecutionMetadata,
	}, nil
}

func GetNodeExecutionInternalData(internalData []byte) (*genModel.NodeExecutionInternalData, error) {
	var nodeExecutionInternalData genModel.NodeExecutionInternalData
	if len(internalData) > 0 {
		err := proto.Unmarshal(internalData, &nodeExecutionInternalData)
		if err != nil {
			return nil, errors.NewFlyteAdminErrorf(codes.Internal, "failed to unmarshal node execution data: %v", err)
		}
	}
	return &nodeExecutionInternalData, nil
}

func handleNodeExecutionInputs(ctx context.Context,
	nodeExecutionModel *models.NodeExecution,
	request *admin.NodeExecutionEventRequest,
	storageClient *storage.DataStore) error {
	if len(nodeExecutionModel.InputURI) > 0 {
		// Inputs are static over the duration of the node execution, no need to update them when they're already set
		return nil
	}
	switch request.GetEvent().GetInputValue().(type) {
	case *event.NodeExecutionEvent_InputUri:
		logger.Debugf(ctx, "saving node execution input URI [%s]", request.GetEvent().GetInputUri())
		nodeExecutionModel.InputURI = request.GetEvent().GetInputUri()
	case *event.NodeExecutionEvent_InputData:
		uri, err := common.OffloadLiteralMap(ctx, storageClient, request.GetEvent().GetInputData(),
			request.GetEvent().GetId().GetExecutionId().GetProject(), request.GetEvent().GetId().GetExecutionId().GetDomain(), request.GetEvent().GetId().GetExecutionId().GetName(),
			request.GetEvent().GetId().GetNodeId(), InputsObjectSuffix)
		if err != nil {
			return err
		}
		logger.Debugf(ctx, "offloaded node execution inputs to [%s]", uri)
		nodeExecutionModel.InputURI = uri.String()
	default:
		logger.Debugf(ctx, "request contained no input data")

	}
	return nil
}

File: flytekit/flytekit/models/node_execution.py
import datetime
import typing
from datetime import timezone as _timezone

import flyteidl.admin.node_execution_pb2 as _node_execution_pb2

from flytekit.models import common as _common_models
from flytekit.models.core import catalog as catalog_models
from flytekit.models.core import compiler as core_compiler_models
from flytekit.models.core import execution as _core_execution
from flytekit.models.core import identifier as _identifier


class WorkflowNodeMetadata(_common_models.FlyteIdlEntity):
    def __init__(self, execution_id: _identifier.WorkflowExecutionIdentifier):
        self._execution_id = execution_id

    @property
    def execution_id(self) -> _identifier.WorkflowExecutionIdentifier:
        return self._execution_id

    def to_flyte_idl(self) -> _node_execution_pb2.WorkflowNodeMetadata:
        return _node_execution_pb2.WorkflowNodeMetadata(
            executionId=self.execution_id.to_flyte_idl(),
        )

    @classmethod
    def from_flyte_idl(cls, p: _node_execution_pb2.WorkflowNodeMetadata) -> "WorkflowNodeMetadata":
        return cls(
            execution_id=_identifier.WorkflowExecutionIdentifier.from_flyte_idl(p.executionId),
        )


class DynamicWorkflowNodeMetadata(_common_models.FlyteIdlEntity):
    def __init__(self, id: _identifier.Identifier, compiled_workflow: core_compiler_models.CompiledWorkflowClosure):
        self._id = id
        self._compiled_workflow = compiled_workflow

    @property
    def id(self) -> _identifier.Identifier:
        return self._id

    @property
    def compiled_workflow(self) -> core_compiler_models.CompiledWorkflowClosure:
        return self._compiled_workflow

    def to_flyte_idl(self) -> _node_execution_pb2.DynamicWorkflowNodeMetadata:
        return _node_execution_pb2.DynamicWorkflowNodeMetadata(
            id=self.id.to_flyte_idl(),
            compiled_workflow=self.compiled_workflow.to_flyte_idl(),
        )

    @classmethod
    def from_flyte_idl(cls, p: _node_execution_pb2.DynamicWorkflowNodeMetadata) -> "DynamicWorkflowNodeMetadata":
        yy = cls(
            id=_identifier.Identifier.from_flyte_idl(p.id),
            compiled_workflow=core_compiler_models.CompiledWorkflowClosure.from_flyte_idl(p.compiled_workflow),
        )
        return yy


class TaskNodeMetadata(_common_models.FlyteIdlEntity):
    def __init__(self, cache_status: int, catalog_key: catalog_models.CatalogMetadata):
        self._cache_status = cache_status
        self._catalog_key = catalog_key

    @property
    def cache_status(self) -> int:
        return self._cache_status

    @property
    def catalog_key(self) -> catalog_models.CatalogMetadata:
        return self._catalog_key

    def to_flyte_idl(self) -> _node_execution_pb2.TaskNodeMetadata:
        return _node_execution_pb2.TaskNodeMetadata(
            cache_status=self.cache_status,
            catalog_key=self.catalog_key.to_flyte_idl(),
        )

    @classmethod
    def from_flyte_idl(cls, p: _node_execution_pb2.TaskNodeMetadata) -> "TaskNodeMetadata":
        return cls(
            cache_status=p.cache_status,
            catalog_key=catalog_models.CatalogMetadata.from_flyte_idl(p.catalog_key),
        )


class NodeExecutionClosure(_common_models.FlyteIdlEntity):
    def __init__(
        self,
        phase,
        started_at,
        duration,
        output_uri=None,
        deck_uri=None,
        error=None,
        workflow_node_metadata: typing.Optional[WorkflowNodeMetadata] = None,
        task_node_metadata: typing.Optional[TaskNodeMetadata] = None,
        created_at: typing.Optional[datetime.datetime] = None,
        updated_at: typing.Optional[datetime.datetime] = None,
    ):
        """
        :param int phase:
        :param datetime.datetime started_at:
        :param datetime.timedelta duration:
        :param Text output_uri:
        :param flytekit.models.core.execution.ExecutionError error:
        """
        self._phase = phase
        self._started_at = started_at
        self._duration = duration
        self._output_uri = output_uri
        self._deck_uri = deck_uri
        self._error = error
        self._workflow_node_metadata = workflow_node_metadata
        self._task_node_metadata = task_node_metadata
        # TODO: Add output_data field as well.
        self._created_at = created_at
        self._updated_at = updated_at

    @property
    def phase(self):
        """
        :rtype: int
        """
        return self._phase

    @property
    def started_at(self):
        """
        :rtype: datetime.datetime
        """
        return self._started_at

    @property
    def duration(self):
        """
        :rtype: datetime.timedelta
        """
        return self._duration

    @property
    def created_at(self) -> typing.Optional[datetime.datetime]:
        return self._created_at

    @property
    def updated_at(self) -> typing.Optional[datetime.datetime]:
        return self._updated_at

    @property
    def output_uri(self):
        """
        :rtype: Text
        """
        return self._output_uri

    @property
    def deck_uri(self):
        """
        :rtype: str
        """
        return self._deck_uri

    @property
    def error(self):
        """
        :rtype: flytekit.models.core.execution.ExecutionError
        """
        return self._error

    @property
    def workflow_node_metadata(self) -> typing.Optional[WorkflowNodeMetadata]:
        return self._workflow_node_metadata

    @property
    def task_node_metadata(self) -> typing.Optional[TaskNodeMetadata]:
        return self._task_node_metadata

    @property
    def target_metadata(self) -> typing.Union[WorkflowNodeMetadata, TaskNodeMetadata]:
        return self.workflow_node_metadata or self.task_node_metadata

    def to_flyte_idl(self):
        """
        :rtype: flyteidl.admin.node_execution_pb2.NodeExecutionClosure
        """
        obj = _node_execution_pb2.NodeExecutionClosure(
            phase=self.phase,
            output_uri=self.output_uri,
            deck_uri=self.deck_uri,
            error=self.error.to_flyte_idl() if self.error is not None else None,
            workflow_node_metadata=self.workflow_node_metadata.to_flyte_idl()
            if self.workflow_node_metadata is not None
            else None,
            task_node_metadata=self.task_node_metadata.to_flyte_idl() if self.task_node_metadata is not None else None,
        )
        obj.started_at.FromDatetime(self.started_at.astimezone(_timezone.utc).replace(tzinfo=None))
        obj.duration.FromTimedelta(self.duration)
        if self.created_at:
            obj.created_at.FromDatetime(self.created_at.astimezone(_timezone.utc).replace(tzinfo=None))
        if self.updated_at:
            obj.updated_at.FromDatetime(self.updated_at.astimezone(_timezone.utc).replace(tzinfo=None))
        return obj

    @classmethod
    def from_flyte_idl(cls, p):
        """
        :param flyteidl.admin.node_execution_pb2.NodeExecutionClosure p:
        :rtype: NodeExecutionClosure
        """
        return cls(
            phase=p.phase,
            output_uri=p.output_uri if p.HasField("output_uri") else None,
            deck_uri=p.deck_uri,
            error=_core_execution.ExecutionError.from_flyte_idl(p.error) if p.HasField("error") else None,
            started_at=p.started_at.ToDatetime().replace(tzinfo=_timezone.utc),
            duration=p.duration.ToTimedelta(),
            workflow_node_metadata=WorkflowNodeMetadata.from_flyte_idl(p.workflow_node_metadata)
            if p.HasField("workflow_node_metadata")
            else None,
            task_node_metadata=TaskNodeMetadata.from_flyte_idl(p.task_node_metadata)
            if p.HasField("task_node_metadata")
            else None,
            created_at=p.created_at.ToDatetime().replace(tzinfo=_timezone.utc) if p.HasField("created_at") else None,
            updated_at=p.updated_at.ToDatetime().replace(tzinfo=_timezone.utc) if p.HasField("updated_at") else None,
        )


class NodeExecutionMetaData(_common_models.FlyteIdlEntity):
    def __init__(self, retry_group: str, is_parent_node: bool, spec_node_id: str):
        self._retry_group = retry_group
        self._is_parent_node = is_parent_node
        self._spec_node_id = spec_node_id

    @property
    def retry_group(self) -> str:
        return self._retry_group

    @property
    def is_parent_node(self) -> bool:
        return self._is_parent_node

    @property
    def spec_node_id(self) -> str:
        return self._spec_node_id

    def to_flyte_idl(self) -> _node_execution_pb2.NodeExecutionMetaData:
        return _node_execution_pb2.NodeExecutionMetaData(
            retry_group=self.retry_group,
            is_parent_node=self.is_parent_node,
            spec_node_id=self.spec_node_id,
        )

    @classmethod
    def from_flyte_idl(cls, p: _node_execution_pb2.NodeExecutionMetaData) -> "NodeExecutionMetaData":
        return cls(
            retry_group=p.retry_group,
            is_parent_node=p.is_parent_node,
            spec_node_id=p.spec_node_id,
        )


class NodeExecution(_common_models.FlyteIdlEntity):
    def __init__(self, id, input_uri, closure, metadata):
        """
        :param flytekit.models.core.identifier.NodeExecutionIdentifier id:
        :param Text input_uri:
        :param NodeExecutionClosure closure:
        :param NodeExecutionMetaData metadata:
        """
        self._id = id
        self._input_uri = input_uri
        self._closure = closure
        self._metadata = metadata

    @property
    def id(self):
        """
        :rtype: flytekit.models.core.identifier.NodeExecutionIdentifier
        """
        return self._id

    @property
    def input_uri(self):
        """
        :rtype: Text
        """
        return self._input_uri

    @property
    def closure(self):
        """
        :rtype: NodeExecutionClosure
        """
        return self._closure

    @property
    def metadata(self) -> NodeExecutionMetaData:
        return self._metadata

    def to_flyte_idl(self) -> _node_execution_pb2.NodeExecution:
        return _node_execution_pb2.NodeExecution(
            id=self.id.to_flyte_idl(),
            input_uri=self.input_uri,
            closure=self.closure.to_flyte_idl(),
            metadata=self.metadata.to_flyte_idl(),
        )

    @classmethod
    def from_flyte_idl(cls, p: _node_execution_pb2.NodeExecution) -> "NodeExecution":
        return cls(
            id=_identifier.NodeExecutionIdentifier.from_flyte_idl(p.id),
            input_uri=p.input_uri,
            closure=NodeExecutionClosure.from_flyte_idl(p.closure),
            metadata=NodeExecutionMetaData.from_flyte_idl(p.metadata),
        )
